crates/transport/src/layers/retry.rs

```diff
@@ -0,0 +1,296 @@
+    requests_enqueued: Arc<AtomicU32>,
+}
+
+impl<S> RetryBackoffService<S> {
+    const fn initial_backoff(&self) -> Duration {
+        Duration::from_millis(self.initial_backoff)
+    }
+}
+
+impl<S> Service<RequestPacket> for RetryBackoffService<S>
+where
+    S: Service<RequestPacket, Response = ResponsePacket, Error = TransportError>
+        + Send
+        + 'static
+        + Clone,
+    S::Future: Send + 'static,
+{
+    type Response = ResponsePacket;
+    type Error = TransportError;
+    type Future = TransportFut<'static>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Our middleware doesn't care about backpressure, so it's ready as long
+        // as the inner service is ready.
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: RequestPacket) -> Self::Future {
+        let inner = self.inner.clone();
+        let this = self.clone();
+        let mut inner = std::mem::replace(&mut self.inner, inner);
+        let mut req = request;
+        Box::pin(async move {
+            let ahead_in_queue = this.requests_enqueued.fetch_add(1, Ordering::SeqCst) as u64;
+            let mut rate_limit_retry_number: u32 = 0;
+            let mut timeout_retries: u32 = 0;
+            let mut batch_success: Vec<Response> = vec![];
+            let mut batch_errs: Vec<Response> = vec![];
+            loop {
+                let err;
+                let res = inner.call(req.clone()).await;
+                match res {
+                    Ok(ResponsePacket::Single(res)) => {
+                        if let Some(e) = res.payload.as_error() {
+                            err = TransportError::ErrorResp(e.clone());
+                        } else {
+                            this.requests_enqueued.fetch_sub(1, Ordering::SeqCst);
+                            return Ok(ResponsePacket::from(vec![res]));
+                        }
+                    }
+                    Ok(ResponsePacket::Batch(batch_res)) => {
+                        for r in batch_res {
+                            if r.is_error() {
+                                batch_errs.push(r);
+                            } else {
+                                batch_success.push(r.clone());
+                                // Remove corresponding request from the batch
+                                req.remove_by_id(&r.id);
+
+                                if req.is_empty() {
+                                    let response_packet =
+                                        ResponsePacket::from(batch_success.clone());
+                                    batch_success.clear();
+                                    batch_errs.clear();
+
+                                    return Ok(response_packet);
+                                }
+                            }
+                        }
+
+                        let e = batch_errs.first().unwrap().payload.as_error().unwrap().clone();
```

Comment:
what does this do?

batch handling looks very complex,

this also only uses the first batch_err, but there could be many